<center><h1><font size="6">DataDive Event Notebook</font></h1>
<h2>Author: Steven Yan</h2></center>

# <a id='0'>Table of Contents</a>

- <a href='#2'>Importing Packages</a>
- <a href='#2'>Importing Packages</a>
    - ACS Datanase
- <a href='#3'>State-Level Databases: New Yrok</a>
    - ACS Database
    - Geodata
    - FCC Database
    - Ookla Database
    - Joining Dataframes
- <a href='#4'>Illinois Datasets


In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# <a id='1'>Importing Packages</a>

In [2]:
import pandas as pd 
import numpy as np
# ! pip install geopandas
import geopandas as gpd 

# Set maximum number of rows and columns that can be viewed
pd.set_option("display.max_columns", 100)
pd.set_option('display.max_rows', 100)


# <a id='2'>Importing Databases</a>

<a href='#0'>Return to Top</a>

Since there seems to be overlap in the terminology, it's best to take a look at how the various databases uses the same nomenclature as the Census Bureau. Generally, state FIPS codes are usually two digits, and county-level FIPS codes are either three or five digits since the state and county FIPS codes can be concatenated to make the five-digit code.


## ACS Database

In [60]:
# Importing ACS database
# data = pd.read_csv('/content/drive/MyDrive/Datasets/Broadband/National/acs_2019.csv.gz', compression ='gzip')
data = pd.read_csv('data/ACS/acs_2019.csv.gz', compression ='gzip')
data.head()

state  county  tract       geoid  f_broadband  f_computer   f_ba  f_black  \
0      1       1  20100  1001020100        0.740       0.815  0.271    0.113   
1      1       1  20200  1001020200        0.669       0.747  0.171    0.582   
2      1       1  20300  1001020300        0.804       0.879  0.184    0.260   
3      1       1  20400  1001020400        0.856       0.926  0.348    0.070   
4      1       1  20500  1001020500        0.890       0.947  0.424    0.192   

   f_hispanic  log_mhi      mhi  population  households  broadband  computer  \
0       0.013    11.01  60208.0      1993.0       709.0       74.0      81.5   
1       0.015    10.69  43958.0      1959.0       688.0       66.9      74.7   
2       0.066    10.92  55345.0      3507.0      1360.0       80.4      87.9   
3       0.011    11.00  59663.0      3878.0      1675.0       85.6      92.6   
4       0.044    11.10  66108.0     10596.0      4483.0       89.0      94.7   

   black  hispanic    mhi.1    ba  den_computers  n_computer  n_broadband  \
0   11.3       1.3  60208.0  27.1          709.0       578.0        525.0   
1   58.2       1.5  43958.0  17.1          688.0       514.0        460.0   
2   26.0       6.6  55345.0  18.4         1360.0      1196.0       1093.0   
3    7.0       1.1  59663.0  34.8         1675.0      1551.0       1434.0   
4   19.2       4.4  66108.0  42.4         4483.0      4246.0       3991.0   

   den_black  n_black  den_hispanic  n_hispanic  den_ba    n_ba  nhh_computer  \
0     1993.0    226.0        1993.0        26.0  1323.0   359.0           578   
1     1959.0   1141.0        1959.0        30.0  1403.0   240.0           514   
2     3507.0    912.0        3507.0       233.0  2505.0   462.0          1196   
3     3878.0    273.0        3878.0        42.0  2856.0   994.0          1551   
4    10596.0   2038.0       10596.0       469.0  6792.0  2878.0          4246   

   nhh_broadband  nhh_computer_any_internet  nhh_computer_and_dialup  \
0            525                        578                        0   
1            460                        514                        0   
2           1093                       1196                        0   
3           1434                       1551                        7   
4           3991                       4246                        0   

   nhh_computer_and_broadband  nhh_computer_no_internet  nhh_no_computer  \
0                         525                        53              131   
1                         460                        54              174   
2                        1068                       128              164   
3                        1429                       115              124   
4                        3991                       255              237   

   den_age  n_children  n_children_computer  n_children_computer_and_dialup  \
0     1993         414                  414                               0   
1     1750         419                  369                               0   
2     3507         807                  723                               0   
3     3878         712                  712                               0   
4    10418        2678                 2678                               0   

   n_children_computer_and_broadband  n_children_computer_no_internet  \
0                                379                               35   
1                                369                                0   
2                                723                                0   
3                                712                                0   
4                               2554                              124   

   n_children_no_computer  
0                       0  
1                      50  
2                      84  
3                       0  
4                       0

In [61]:
# Get the counts of the values in tract column
data.tract.value_counts().head()

950100    382
950200    343
100       334
200       331
300       328
Name: tract, dtype: int64

In [62]:
# Get the counts of the values in geoid column
data.geoid.value_counts().head()

1001020100     1
37063002019    1
37063002025    1
37063002024    1
37063002023    1
Name: geoid, dtype: int64

In [63]:
data.columns

Index(['state', 'county', 'tract', 'geoid', 'f_broadband', 'f_computer',
       'f_ba', 'f_black', 'f_hispanic', 'log_mhi', 'mhi', 'population',
       'households', 'broadband', 'computer', 'black', 'hispanic', 'mhi.1',
       'ba', 'den_computers', 'n_computer', 'n_broadband', 'den_black',
       'n_black', 'den_hispanic', 'n_hispanic', 'den_ba', 'n_ba',
       'nhh_computer', 'nhh_broadband', 'nhh_computer_any_internet',
       'nhh_computer_and_dialup', 'nhh_computer_and_broadband',
       'nhh_computer_no_internet', 'nhh_no_computer', 'den_age', 'n_children',
       'n_children_computer', 'n_children_computer_and_dialup',
       'n_children_computer_and_broadband', 'n_children_computer_no_internet',
       'n_children_no_computer'],
      dtype='object')

The columns for the database are:

- `state`
- `county`
- `tract`
- `geoid`
- `f_broadband`:  Broadband Access, 
- `f_computer`, `f_ba`, `f_black`, `f_hispanic`
- `log_mhi`:  Log Median Household Income
- `mhi`: Median Household Income
- `population`
- `households`
- `broadband`
- `computer`
- `black`
- `hispanic`
- `mhi.1`
- `ba`
- `den_computers`, `den_age`, `den_black`, `den_hispanic`, `den_ba`
- `n_computer`, `n_broadband`, `n_black`, `n_hispanic`. `n_ba`, `n_children`, `n_children_computer`, `n_children_computer_and_dialup`, `n_children_computer_and_broadband`, `n_children_computer_no_internet`, `n_children_no_computer`
- `nhh_computer`, , `nhh_computer_any_internet`, `nhh_computer_and_dialup`, `nhh_computer_and_broadband`, `nhh_computer_no_internet`, `nhh_no_computer`
- `nhh_no_broadband` = `households` - `nhh_broadband`
- `n_children_no_computer_and_broadband` = `n_children` - `n_children_computer_and_broadband`

I haven't been able to entirely figure out most of the columns mean.  n is either no or number.  nhh is no household?





# <a id='2'>State-level databases</a>

<a href='#0'>Return to Top</a>

## Geodata


In [65]:
# import shapefile into dataframe
# ny_shp = gpd.read_file("/content/drive/MyDrive/Datasets/Broadband/NY/ny_spdf/ny_spdf.shp")
ny_shp = gpd.read_file("data/NY/ny_spdf/ny_spdf.shp")
ny_shp.head()

STATEFP COUNTYFP TRACTCE              AFFGEOID        GEOID     NAME LSAD  \
0      36      071  000502  1400000US36071000502  36071000502     5.02   CT   
1      36      103  135208  1400000US36103135208  36103135208  1352.08   CT   
2      36      047  057800  1400000US36047057800  36047057800      578   CT   
3      36      047  058900  1400000US36047058900  36047058900      589   CT   
4      36      055  013204  1400000US36055013204  36055013204   132.04   CT   

        ALAND    AWATER                                           geometry  
0    967431.0  969216.0  POLYGON ((-74.02226 41.49281, -74.02180 41.496...  
1   2287077.0       0.0  POLYGON ((-73.28263 40.83063, -73.28157 40.832...  
2    172233.0       0.0  POLYGON ((-73.95398 40.60140, -73.95304 40.601...  
3    424025.0   38353.0  POLYGON ((-73.94605 40.72926, -73.94419 40.729...  
4  28207247.0   93541.0  POLYGON ((-77.66835 43.02829, -77.66806 43.029...

In [8]:
# get overview of dataframe
ny_shp.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
RangeIndex: 4906 entries, 0 to 4905
Data columns (total 10 columns):
 #   Column    Non-Null Count  Dtype   
---  ------    --------------  -----   
 0   STATEFP   4906 non-null   object  
 1   COUNTYFP  4906 non-null   object  
 2   TRACTCE   4906 non-null   object  
 3   AFFGEOID  4906 non-null   object  
 4   GEOID     4906 non-null   object  
 5   NAME      4906 non-null   object  
 6   LSAD      4906 non-null   object  
 7   ALAND     4906 non-null   float64 
 8   AWATER    4906 non-null   float64 
 9   geometry  4906 non-null   geometry
dtypes: float64(2), geometry(1), object(7)
memory usage: 383.4+ KB


In [9]:
# get number of different county codes
ny_shp.COUNTYFP.nunique()

62

In [10]:
# Get maximum value for county code
ny_shp.COUNTYFP.max()

'123'

In [11]:
# get some information about GEOID
print("Number of unique values: ", ny_shp.GEOID.nunique())
print("Minimum value: ", ny_shp.GEOID.min())
print("Maximum value: ", ny_shp.GEOID.max())

Number of unique values:  4906
Minimum value:  36001000100
Maximum value:  36123150500


In [12]:
# get some information about TRACTCE
print("Number of unique values: ", ny_shp.TRACTCE.nunique())
print("Minimum value: ", ny_shp.TRACTCE.min())
print("Maximum value: ", ny_shp.TRACTCE.max())

Number of unique values:  2697
Minimum value:  000100
Maximum value:  990100


In [13]:
# get value counts of TRACTCE
ny_df = ny_shp.TRACTCE.value_counts().rename_axis('unique_values').reset_index(name='counts')
ny_df.head(10)

unique_values  counts
0        000600      12
1        000700      12
2        000200      12
3        010400      11
4        010200      11
5        013800      11
6        011000      11
7        001600      11
8        000100      11
9        011800      10

In [14]:
ny_000600 = ny_shp[ny_shp['TRACTCE']=='000600']
ny_000600.head()

STATEFP COUNTYFP TRACTCE              AFFGEOID        GEOID NAME LSAD  \
135       36      001  000600  1400000US36001000600  36001000600    6   CT   
830       36      007  000600  1400000US36007000600  36007000600    6   CT   
1286      36      061  000600  1400000US36061000600  36061000600    6   CT   
1359      36      015  000600  1400000US36015000600  36015000600    6   CT   
1573      36      029  000600  1400000US36029000600  36029000600    6   CT   

          ALAND    AWATER                                           geometry  
135    506881.0       0.0  POLYGON ((-73.77924 42.67116, -73.77148 42.666...  
830   1551772.0   12657.0  POLYGON ((-75.89499 42.10717, -75.89437 42.108...  
1286   240406.0  176018.0  POLYGON ((-73.99256 40.71439, -73.99022 40.714...  
1359  1117293.0   39031.0  POLYGON ((-76.82248 42.09642, -76.81913 42.097...  
1573  1163904.0       0.0  POLYGON ((-78.82385 42.83482, -78.82382 42.837...

In [66]:
# sort dataframe by GEOID values
df_1 = ny_shp.sort_values(by=['GEOID'])
df_1.head()

STATEFP COUNTYFP TRACTCE              AFFGEOID        GEOID  NAME LSAD  \
539       36      001  000100  1400000US36001000100  36001000100     1   CT   
2748      36      001  000200  1400000US36001000200  36001000200     2   CT   
3007      36      001  000300  1400000US36001000300  36001000300     3   CT   
3515      36      001  000401  1400000US36001000401  36001000401  4.01   CT   
2751      36      001  000403  1400000US36001000403  36001000403  4.03   CT   

          ALAND    AWATER                                           geometry  
539   2367456.0  245021.0  POLYGON ((-73.74506 42.67228, -73.74374 42.677...  
2748  2065161.0       0.0  POLYGON ((-73.75993 42.65997, -73.76156 42.660...  
3007  5758852.0   55326.0  POLYGON ((-73.83021 42.69418, -73.82785 42.696...  
3515  9065815.0   86327.0  POLYGON ((-73.89421 42.70977, -73.89668 42.711...  
2751  3138700.0       0.0  POLYGON ((-73.82108 42.67857, -73.81916 42.680...

## Synopsis of NY Shapefile:

There are 4906 rows in the shapefile, as well 
as 4906 unique rows of GEOID's. The state FIPS code is 36, and there are 62 county FIPS code in NY.  The AFFGEOID seems to be the GEOID affixed with 1400000US in front of each GEOID.  There are values given for land area, water area, and most importantly the geometry of the tract boundaries.

Originally, I thought that TRACTCEs needed to be concatentated together as they each represented a unique value, but instead they represent a 6-digit tract FIPS code.  

Each GEOID consists of the state FIPS code, its county FIPS code, and then the tract FIPS code, which apparently in itself is not unique and can be repeated, so 2 + 3 + 6 digits in total for the three components or an 11-digit tract FIPS code that is unique for every tract.

## ACS Database

In [67]:
# import database
# ny_1 = pd.read_csv('/content/drive/MyDrive/Datasets/Broadband/NY/acs_2019_NY.csv')
ny_1 = pd.read_csv('data/NY/acs_2019_NY.csv')
ny_1.head()

state  county  tract        geoid  f_broadband  f_computer   f_ba  f_black  \
0     36       1    100  36001000100        0.836       0.896  0.137    0.740   
1     36       1    200  36001000200        0.767       0.903  0.225    0.845   
2     36       1    300  36001000300        0.819       0.881  0.289    0.435   
3     36       1    401  36001000401        0.837       0.857  0.476    0.076   
4     36       1    403  36001000403        0.817       0.861  0.591    0.129   

   f_hispanic  log_mhi      mhi  population  households  broadband  computer  \
0       0.128    10.39  32389.0      2035.0       800.0       83.6      89.6   
1       0.043    10.23  27714.0      4793.0      2213.0       76.7      90.3   
2       0.121    10.72  45272.0      6147.0      2362.0       81.9      88.1   
3       0.046    11.22  74274.0      2362.0       997.0       83.7      85.7   
4       0.058    11.22  74426.0      4253.0      2186.0       81.7      86.1   

   black  hispanic    mhi.1    ba  den_computers  n_computer  n_broadband  \
0   74.0      12.8  32389.0  13.7          800.0       717.0        669.0   
1   84.5       4.3  27714.0  22.5         2213.0      1999.0       1697.0   
2   43.5      12.1  45272.0  28.9         2362.0      2081.0       1935.0   
3    7.6       4.6  74274.0  47.6          997.0       854.0        834.0   
4   12.9       5.8  74426.0  59.1         2186.0      1882.0       1786.0   

   den_black  n_black  den_hispanic  n_hispanic  den_ba    n_ba  nhh_computer  \
0     2035.0   1506.0        2035.0       261.0  1283.0   176.0           717   
1     4793.0   4050.0        4793.0       206.0  2710.0   609.0          1999   
2     6147.0   2674.0        6147.0       742.0  3725.0  1078.0          2081   
3     2362.0    180.0        2362.0       109.0  2169.0  1032.0           854   
4     4253.0    548.0        4253.0       246.0  3220.0  1904.0          1882   

   nhh_broadband  nhh_computer_any_internet  nhh_computer_and_dialup  \
0            669                        717                        0   
1           1697                       1999                        0   
2           1935                       2081                       11   
3            834                        854                        0   
4           1786                       1882                        0   

   nhh_computer_and_broadband  nhh_computer_no_internet  nhh_no_computer  \
0                         669                        48               83   
1                        1685                       314              214   
2                        1917                       153              281   
3                         826                        28              143   
4                        1733                       149              304   

   den_age  n_children  n_children_computer  n_children_computer_and_dialup  \
0     2022         532                  520                               0   
1     4711        1274                 1221                               0   
2     5925        1729                 1717                               0   
3     1790         171                  171                               0   
4     4240         462                  462                               0   

   n_children_computer_and_broadband  n_children_computer_no_internet  \
0                                511                                9   
1                               1138                               83   
2                               1630                               87   
3                                171                                0   
4                                462                                0   

   n_children_no_computer  state_lkp  
0                      12         36  
1                      53         36  
2                      12         36  
3                       0         36  
4                       0         36

In [17]:
ny_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4918 entries, 0 to 4917
Data columns (total 43 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   state                              4918 non-null   int64  
 1   county                             4918 non-null   int64  
 2   tract                              4918 non-null   int64  
 3   geoid                              4918 non-null   int64  
 4   f_broadband                        4918 non-null   float64
 5   f_computer                         4918 non-null   float64
 6   f_ba                               4918 non-null   float64
 7   f_black                            4918 non-null   float64
 8   f_hispanic                         4918 non-null   float64
 9   log_mhi                            4785 non-null   float64
 10  mhi                                4918 non-null   float64
 11  population                         4918 non-null   float

In [18]:
# check for any null values
ny_1.isna().sum()

state                                  0
county                                 0
tract                                  0
geoid                                  0
f_broadband                            0
f_computer                             0
f_ba                                   0
f_black                                0
f_hispanic                             0
log_mhi                              133
mhi                                    0
population                             0
households                             0
broadband                              0
computer                               0
black                                  0
hispanic                               0
mhi.1                                  0
ba                                     0
den_computers                          0
n_computer                             0
n_broadband                            0
den_black                              0
n_black                                0
den_hispanic    

In [19]:
# get some information about tract
print("Number of unique values: ", ny_1.tract.nunique())
print("Minimum value: ", ny_1.tract.min())
print("Maximum value: ", ny_1.tract.max())

Number of unique values:  2701
Minimum value:  100
Maximum value:  990400


In [20]:
# get some information about geoid
print("Number of unique values: ", ny_1.geoid.nunique())
print("Minimum value: ", ny_1.geoid.min())
print("Maximum value: ", ny_1.geoid.max())


Number of unique values:  4918
Minimum value:  36001000100
Maximum value:  36123150500


In [68]:
ny_1['geoid'] = ny_1['geoid'].apply(lambda x: str(x))

In [69]:
joined_df = df_1.merge(ny_1, how='left', left_on='GEOID', right_on='geoid')
joined_df.head()

STATEFP COUNTYFP TRACTCE              AFFGEOID        GEOID  NAME LSAD  \
0      36      001  000100  1400000US36001000100  36001000100     1   CT   
1      36      001  000200  1400000US36001000200  36001000200     2   CT   
2      36      001  000300  1400000US36001000300  36001000300     3   CT   
3      36      001  000401  1400000US36001000401  36001000401  4.01   CT   
4      36      001  000403  1400000US36001000403  36001000403  4.03   CT   

       ALAND    AWATER                                           geometry  \
0  2367456.0  245021.0  POLYGON ((-73.74506 42.67228, -73.74374 42.677...   
1  2065161.0       0.0  POLYGON ((-73.75993 42.65997, -73.76156 42.660...   
2  5758852.0   55326.0  POLYGON ((-73.83021 42.69418, -73.82785 42.696...   
3  9065815.0   86327.0  POLYGON ((-73.89421 42.70977, -73.89668 42.711...   
4  3138700.0       0.0  POLYGON ((-73.82108 42.67857, -73.81916 42.680...   

   state  county  tract        geoid  f_broadband  f_computer   f_ba  f_black  \
0     36       1    100  36001000100        0.836       0.896  0.137    0.740   
1     36       1    200  36001000200        0.767       0.903  0.225    0.845   
2     36       1    300  36001000300        0.819       0.881  0.289    0.435   
3     36       1    401  36001000401        0.837       0.857  0.476    0.076   
4     36       1    403  36001000403        0.817       0.861  0.591    0.129   

   f_hispanic  log_mhi      mhi  population  households  broadband  computer  \
0       0.128    10.39  32389.0      2035.0       800.0       83.6      89.6   
1       0.043    10.23  27714.0      4793.0      2213.0       76.7      90.3   
2       0.121    10.72  45272.0      6147.0      2362.0       81.9      88.1   
3       0.046    11.22  74274.0      2362.0       997.0       83.7      85.7   
4       0.058    11.22  74426.0      4253.0      2186.0       81.7      86.1   

   black  hispanic    mhi.1    ba  den_computers  n_computer  n_broadband  \
0   74.0      12.8  32389.0  13.7          800.0       717.0        669.0   
1   84.5       4.3  27714.0  22.5         2213.0      1999.0       1697.0   
2   43.5      12.1  45272.0  28.9         2362.0      2081.0       1935.0   
3    7.6       4.6  74274.0  47.6          997.0       854.0        834.0   
4   12.9       5.8  74426.0  59.1         2186.0      1882.0       1786.0   

   den_black  n_black  den_hispanic  n_hispanic  den_ba    n_ba  nhh_computer  \
0     2035.0   1506.0        2035.0       261.0  1283.0   176.0           717   
1     4793.0   4050.0        4793.0       206.0  2710.0   609.0          1999   
2     6147.0   2674.0        6147.0       742.0  3725.0  1078.0          2081   
3     2362.0    180.0        2362.0       109.0  2169.0  1032.0           854   
4     4253.0    548.0        4253.0       246.0  3220.0  1904.0          1882   

   nhh_broadband  nhh_computer_any_internet  nhh_computer_and_dialup  \
0            669                        717                        0   
1           1697                       1999                        0   
2           1935                       2081                       11   
3            834                        854                        0   
4           1786                       1882                        0   

   nhh_computer_and_broadband  nhh_computer_no_internet  nhh_no_computer  \
0                         669                        48               83   
1                        1685                       314              214   
2                        1917                       153              281   
3                         826                        28              143   
4                        1733                       149              304   

   den_age  n_children  n_children_computer  n_children_computer_and_dialup  \
0     2022         532                  520                               0   
1     4711        1274                 1221                               0   
2     5925        1729               

In [70]:
joined_df.drop(columns=['STATEFP', 'COUNTYFP', 'TRACTCE', 'AFFGEOID', 'NAME', 'LSAD', 'ALAND', 'AWATER', 'state', 'county', 'tract', 'geoid', 'state_lkp'], inplace=True)

In [71]:
joined_df.head()

GEOID                                           geometry  \
0  36001000100  POLYGON ((-73.74506 42.67228, -73.74374 42.677...   
1  36001000200  POLYGON ((-73.75993 42.65997, -73.76156 42.660...   
2  36001000300  POLYGON ((-73.83021 42.69418, -73.82785 42.696...   
3  36001000401  POLYGON ((-73.89421 42.70977, -73.89668 42.711...   
4  36001000403  POLYGON ((-73.82108 42.67857, -73.81916 42.680...   

   f_broadband  f_computer   f_ba  f_black  f_hispanic  log_mhi      mhi  \
0        0.836       0.896  0.137    0.740       0.128    10.39  32389.0   
1        0.767       0.903  0.225    0.845       0.043    10.23  27714.0   
2        0.819       0.881  0.289    0.435       0.121    10.72  45272.0   
3        0.837       0.857  0.476    0.076       0.046    11.22  74274.0   
4        0.817       0.861  0.591    0.129       0.058    11.22  74426.0   

   population  households  broadband  computer  black  hispanic    mhi.1  \
0      2035.0       800.0       83.6      89.6   74.0      12.8  32389.0   
1      4793.0      2213.0       76.7      90.3   84.5       4.3  27714.0   
2      6147.0      2362.0       81.9      88.1   43.5      12.1  45272.0   
3      2362.0       997.0       83.7      85.7    7.6       4.6  74274.0   
4      4253.0      2186.0       81.7      86.1   12.9       5.8  74426.0   

     ba  den_computers  n_computer  n_broadband  den_black  n_black  \
0  13.7          800.0       717.0        669.0     2035.0   1506.0   
1  22.5         2213.0      1999.0       1697.0     4793.0   4050.0   
2  28.9         2362.0      2081.0       1935.0     6147.0   2674.0   
3  47.6          997.0       854.0        834.0     2362.0    180.0   
4  59.1         2186.0      1882.0       1786.0     4253.0    548.0   

   den_hispanic  n_hispanic  den_ba    n_ba  nhh_computer  nhh_broadband  \
0        2035.0       261.0  1283.0   176.0           717            669   
1        4793.0       206.0  2710.0   609.0          1999           1697   
2        6147.0       742.0  3725.0  1078.0          2081           1935   
3        2362.0       109.0  2169.0  1032.0           854            834   
4        4253.0       246.0  3220.0  1904.0          1882           1786   

   nhh_computer_any_internet  nhh_computer_and_dialup  \
0                        717                        0   
1                       1999                        0   
2                       2081                       11   
3                        854                        0   
4                       1882                        0   

   nhh_computer_and_broadband  nhh_computer_no_internet  nhh_no_computer  \
0                         669                        48               83   
1                        1685                       314              214   
2                        1917                       153              281   
3                         826                        28              143   
4                        1733                       149              304   

   den_age  n_children  n_children_computer  n_children_computer_and_dialup  \
0     2022         532                  520                               0   
1     4711        1274                 1221                               0   
2     5925        1729                 1717                               0   
3     1790         171                  171                               0   
4     4240         462                  462                               0   

   n_children_computer_and_broadband  n_children_computer_no_internet  \
0                                511                                9   
1                               1138                               83   
2                               1630                               87   
3                                171                                0   
4                                462                                0   

   n_children_no_computer  
0                      12  
1                      53 

In [72]:
type(joined_df)

geopandas.geodataframe.GeoDataFrame

In [26]:
joined_df.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
Int64Index: 4906 entries, 0 to 4905
Data columns (total 40 columns):
 #   Column                             Non-Null Count  Dtype   
---  ------                             --------------  -----   
 0   GEOID                              4906 non-null   object  
 1   geometry                           4906 non-null   geometry
 2   f_broadband                        4906 non-null   float64 
 3   f_computer                         4906 non-null   float64 
 4   f_ba                               4906 non-null   float64 
 5   f_black                            4906 non-null   float64 
 6   f_hispanic                         4906 non-null   float64 
 7   log_mhi                            4785 non-null   float64 
 8   mhi                                4906 non-null   float64 
 9   population                         4906 non-null   float64 
 10  households                         4906 non-null   float64 
 11  broadband                          

## FCC Database

The columns are as follows:

- `tract`: 11-digit tract FIPS code
- `max_dn`: average maximum advertised downstream speed offered by provider in census tract by block
- `max_up`: average maximum advertised upstream speed offered by provider in census tract by block
- `dn10`:	average count of providers offering advertised downstream greater than 10 by block
- `dn100`: average count of providers offering advertised downstream greater than 100 by block
- `dn250`: average count of providers offering advertised downstream greater than 250 by block
- `fiber_100u`: average count of providers offiering fiber by block


In [73]:
# importing FCC database
# ny_2 = pd.read_csv('/content/drive/MyDrive/Datasets/Broadband/NY/fcc_477_census_tract_NY.csv')
ny_2 = pd.read_csv('data/NY/fcc_477_census_tract_NY.csv')

ny_2.head()

tract      max_dn     max_up      dn10     dn100     dn250  \
0  36001000100  711.250000  50.722222  1.875000  0.777778  0.777778   
1  36001000200  940.000000  35.000000  2.529412  1.011765  1.000000   
2  36001000300  928.040541  98.195946  2.290541  1.060811  1.060811   
3  36001000401  928.975904  65.168675  2.156627  1.024096  1.024096   
4  36001000403  940.000000  43.203883  2.310680  1.009709  1.009709   

   fiber_100u  state  
0    0.027778     36  
1    0.000000     36  
2    0.074324     36  
3    0.036145     36  
4    0.009709     36

In [28]:
# check for any null values in dataframe
ny_2.isna().sum()

tract         0
max_dn        0
max_up        0
dn10          0
dn100         0
dn250         0
fiber_100u    0
state         0
dtype: int64

In [29]:
# show value counts of tract column
ny_2.tract.value_counts()

36029003800    1
36093020800    1
36029004402    1
36027110003    1
36029016700    1
              ..
36029011500    1
36047082800    1
36093021200    1
36005008700    1
36019101700    1
Name: tract, Length: 4900, dtype: int64

In [30]:
# find unique values of tract column
ny_2.tract.nunique()

4900

In [74]:
# change tract column to string type for joining
ny_2['tract'] = ny_2['tract'].apply(lambda x: str(x))
type(ny_2['tract'][0])

str

In [75]:
# merge two database on the unique 11-digit FIPS code
df_2 = joined_df.merge(ny_2, how='left', left_on='GEOID', right_on='tract')
df_2.head()

GEOID                                           geometry  \
0  36001000100  POLYGON ((-73.74506 42.67228, -73.74374 42.677...   
1  36001000200  POLYGON ((-73.75993 42.65997, -73.76156 42.660...   
2  36001000300  POLYGON ((-73.83021 42.69418, -73.82785 42.696...   
3  36001000401  POLYGON ((-73.89421 42.70977, -73.89668 42.711...   
4  36001000403  POLYGON ((-73.82108 42.67857, -73.81916 42.680...   

   f_broadband  f_computer   f_ba  f_black  f_hispanic  log_mhi      mhi  \
0        0.836       0.896  0.137    0.740       0.128    10.39  32389.0   
1        0.767       0.903  0.225    0.845       0.043    10.23  27714.0   
2        0.819       0.881  0.289    0.435       0.121    10.72  45272.0   
3        0.837       0.857  0.476    0.076       0.046    11.22  74274.0   
4        0.817       0.861  0.591    0.129       0.058    11.22  74426.0   

   population  households  broadband  computer  black  hispanic    mhi.1  \
0      2035.0       800.0       83.6      89.6   74.0      12.8  32389.0   
1      4793.0      2213.0       76.7      90.3   84.5       4.3  27714.0   
2      6147.0      2362.0       81.9      88.1   43.5      12.1  45272.0   
3      2362.0       997.0       83.7      85.7    7.6       4.6  74274.0   
4      4253.0      2186.0       81.7      86.1   12.9       5.8  74426.0   

     ba  den_computers  n_computer  n_broadband  den_black  n_black  \
0  13.7          800.0       717.0        669.0     2035.0   1506.0   
1  22.5         2213.0      1999.0       1697.0     4793.0   4050.0   
2  28.9         2362.0      2081.0       1935.0     6147.0   2674.0   
3  47.6          997.0       854.0        834.0     2362.0    180.0   
4  59.1         2186.0      1882.0       1786.0     4253.0    548.0   

   den_hispanic  n_hispanic  den_ba    n_ba  nhh_computer  nhh_broadband  \
0        2035.0       261.0  1283.0   176.0           717            669   
1        4793.0       206.0  2710.0   609.0          1999           1697   
2        6147.0       742.0  3725.0  1078.0          2081           1935   
3        2362.0       109.0  2169.0  1032.0           854            834   
4        4253.0       246.0  3220.0  1904.0          1882           1786   

   nhh_computer_any_internet  nhh_computer_and_dialup  \
0                        717                        0   
1                       1999                        0   
2                       2081                       11   
3                        854                        0   
4                       1882                        0   

   nhh_computer_and_broadband  nhh_computer_no_internet  nhh_no_computer  \
0                         669                        48               83   
1                        1685                       314              214   
2                        1917                       153              281   
3                         826                        28              143   
4                        1733                       149              304   

   den_age  n_children  n_children_computer  n_children_computer_and_dialup  \
0     2022         532                  520                               0   
1     4711        1274                 1221                               0   
2     5925        1729                 1717                               0   
3     1790         171                  171                               0   
4     4240         462                  462                               0   

   n_children_computer_and_broadband  n_children_computer_no_internet  \
0                                511                                9   
1                               1138                               83   
2                               1630                               87   
3                                171                                0   
4                                462                                0   

   n_children_no_computer        tract      max_dn     max_up      dn10  \
0      

In [76]:
df_2.drop(columns=['tract', 'state'], inplace=True)

In [77]:
df_2.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
Int64Index: 4906 entries, 0 to 4905
Data columns (total 46 columns):
 #   Column                             Non-Null Count  Dtype   
---  ------                             --------------  -----   
 0   GEOID                              4906 non-null   object  
 1   geometry                           4906 non-null   geometry
 2   f_broadband                        4906 non-null   float64 
 3   f_computer                         4906 non-null   float64 
 4   f_ba                               4906 non-null   float64 
 5   f_black                            4906 non-null   float64 
 6   f_hispanic                         4906 non-null   float64 
 7   log_mhi                            4785 non-null   float64 
 8   mhi                                4906 non-null   float64 
 9   population                         4906 non-null   float64 
 10  households                         4906 non-null   float64 
 11  broadband                          

In [35]:
type(df_2)

geopandas.geodataframe.GeoDataFrame

## Synopsis of FCC Database

FCC has 4900 rows of data, and the shapefile had 4906 rows of each of the tracts, and every row has a unique tract value. Here `tract` is synonymous with `GEOID`. It would seem we are missing data for a few GEOIDs, but that remains to be seen how our analysis will be affected.

## Ookla Database

The column keys are:
- `avg_d_kbps`, `int`, average download speed of all tests performed in the tile, represented in kilobits per second.
- `avg_u_kbps`, `int`, 	average upload speed of all tests performed in the tile, represented in kilobits per second.
- `avg_lat_ms`, `int`, average latency of all tests performed in the tile, represented in milliseconds
- `tests`, `int`, number of tests taken in the tile.
- `devices`, `int`, number of unique devices contributing tests in the tile.
- `quadkey`: quadkey representing the tile.



In [78]:
# import Ookla database
ny_3 = pd.read_csv('data/NY/ookla_combined_ny.csv')
# ny_3 = pd.read_csv('/content/drive/MyDrive/Datasets/Broadband/NY/ookla_combined_ny.csv')
ny_3.head()

quadkey  avg_d_kbps  avg_u_kbps  avg_lat_ms  tests  devices   type  \
0  302332203102320      208526        7394          14     50        2  fixed   
1  302231312100203      153742       97895           6    134       28  fixed   
2  320101103112201      202082       55540          15    222       56  fixed   
3  302322000030131      176743       14699          20     18        3  fixed   
4  302233111212211      370517      311427           4      2        1  fixed   

  quarter  year  STATEFP  COUNTYFP  TRACTCE              AFFGEOID  \
0      Q1  2021       36       103   170202  1400000US36103170202   
1      Q1  2021       36        63    24502  1400000US36063024502   
2      Q1  2021       36        47    92400  1400000US36047092400   
3      Q1  2021       36        29    14702  1400000US36029014702   
4      Q1  2021       36        29     7302  1400000US36029007302   

         GEOID     NAME LSAD     ALAND     AWATER  \
0  36103170202  1702.02   CT  41367468  100616534   
1  36063024502   245.02   CT  77726398   10786358   
2  36047092400   924.00   CT    251687          0   
3  36029014702   147.02   CT  31997799     308205   
4  36029007302    73.02   CT  36949030    5791136   

                                                tile  
0  list(c(-72.57568359375, -72.5701904296875, -72...  
1  list(c(-79.2718505859375, -79.266357421875, -7...  
2  list(c(-73.9105224609375, -73.905029296875, -7...  
3  list(c(-78.6236572265625, -78.6181640625, -78....  
4  list(c(-78.9971923828125, -78.99169921875, -78...

In [49]:
# get general attributes of dataframe
ny_3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 444771 entries, 0 to 444770
Data columns (total 19 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   quadkey     444771 non-null  int64  
 1   avg_d_kbps  444771 non-null  int64  
 2   avg_u_kbps  444771 non-null  int64  
 3   avg_lat_ms  444771 non-null  int64  
 4   tests       444771 non-null  int64  
 5   devices     444771 non-null  int64  
 6   type        444771 non-null  object 
 7   quarter     444771 non-null  object 
 8   year        444771 non-null  int64  
 9   STATEFP     444771 non-null  int64  
 10  COUNTYFP    444771 non-null  int64  
 11  TRACTCE     444771 non-null  int64  
 12  AFFGEOID    444771 non-null  object 
 13  GEOID       444771 non-null  int64  
 14  NAME        444771 non-null  float64
 15  LSAD        444771 non-null  object 
 16  ALAND       444771 non-null  int64  
 17  AWATER      444771 non-null  int64  
 18  tile        444771 non-null  object 
dtypes:

In [50]:
# get unique values of GEOID
ny_3.GEOID.nunique()

3891

The number of rows are significantly higher than the other databases, so there are multiple instances of each tract area.  There are also only 3891 different tract areas represented, which is more than 1000 less than the total off 4906 in the shapefiles.

In [53]:
# check for nnull valaues
ny_3.isna().sum()

quadkey       0
avg_d_kbps    0
avg_u_kbps    0
avg_lat_ms    0
tests         0
devices       0
type          0
quarter       0
year          0
STATEFP       0
COUNTYFP      0
TRACTCE       0
AFFGEOID      0
GEOID         0
NAME          0
LSAD          0
ALAND         0
AWATER        0
tile          0
dtype: int64

In [54]:
# get unique values of tile column
ny_3.tile.nunique()

234446

In [55]:
# get unique values of quadkeys column
ny_3.quadkey.nunique()

234446

## Attempting to address quadkeys and tiles

In [21]:
fixed_ookla_shp = gpd.read_file('data/z-archive/Ookla/ookla-data/2019-01-01_performance_fixed_tiles/gps_fixed_tiles.shp')
fixed_ookla_shp

quadkey  avg_d_kbps  avg_u_kbps  avg_lat_ms  tests  devices  \
0        0231113112003202       66216       12490          13     28        4   
1        1322111021111001      102598       37356          13     15        4   
2        3112203030003110       24686       18736          18    162      106   
3        0320000130321312       17674       13989          78    364        4   
4        0320001332313103      441192      218955          22     14        1   
...                   ...         ...         ...         ...    ...      ...   
4877031  1322031000321100       55452       55740           6      1        1   
4877032  1220110033212313       37277       10907          20      3        1   
4877033  1221210121203010         661        1088          64      1        1   
4877034  1202322013330233       35493       14623          20      2        1   
4877035  1220111202123221        4607         517          29      1        1   

                                                  geometry  
0        POLYGON ((-90.65918 38.49229, -90.65369 38.492...  
1        POLYGON ((110.35217 21.28937, 110.35767 21.289...  
2        POLYGON ((138.59253 -34.92197, 138.59802 -34.9...  
3        POLYGON ((-87.63794 40.22502, -87.63245 40.225...  
4        POLYGON ((-84.74304 38.92096, -84.73755 38.920...  
...                                                    ...  
4877031  POLYGON ((98.67920 16.38339, 98.68469 16.38339...  
4877032  POLYGON ((18.05603 39.98975, 18.06152 39.98975...  
4877033  POLYGON ((29.93774 31.16581, 29.94324 31.16581...  
4877034  POLYGON ((12.58484 42.58949, 12.59033 42.58949...  
4877035  POLYGON ((19.91272 39.51252, 19.91821 39.51252...  

[4877036 rows x 7 columns]

In [22]:
mobile_ookla_shp = gpd.read_file('data/z-archive/Ookla/ookla-data/2019-01-01_performance_mobile_tiles/gps_mobile_tiles.shp')
mobile_ookla_shp

quadkey  avg_d_kbps  avg_u_kbps  avg_lat_ms  tests  devices  \
0        1202130120303121       12503        5090          43      1        1   
1        0322113021201023       16109       11204          36     98       49   
2        3100130002212100       10325        7378          46     94       34   
3        1231213031333131        7187        2051          50    211      104   
4        0230102031111210       76282       20332          27      2        2   
...                   ...         ...         ...         ...    ...      ...   
3231240  1321022032310213        9746        5092          34      1        1   
3231241  1202033232332030        4894         870          33      1        1   
3231242  0302231230311200        8218        8480          26      1        1   
3231243  3100131231013021       15525        3846          43      1        1   
3231244  1221300101221111       30911       21403          16      1        1   

                                                  geometry  
0        POLYGON ((18.52844 51.91039, 18.53394 51.91039...  
1        POLYGON ((-69.91150 18.46398, -69.90601 18.463...  
2        POLYGON ((106.98486 -6.18425, 106.99036 -6.184...  
3        POLYGON ((77.33826 28.64239, 77.34375 28.64239...  
4        POLYGON ((-122.37671 38.25544, -122.37122 38.2...  
...                                                    ...  
3231240  POLYGON ((113.48328 33.26166, 113.48877 33.261...  
3231241  POLYGON ((9.41528 48.94415, 9.42078 48.94415, ...  
3231242  POLYGON ((-80.55176 43.43697, -80.54626 43.436...  
3231243  POLYGON ((110.87952 -7.76542, 110.88501 -7.765...  
3231244  POLYGON ((35.59021 31.72817, 35.59570 31.72817...  

[3231245 rows x 7 columns]

In [29]:
us_tract_shp = gpd.read_file("data/z-archive/Census Tract Shapefiles/cb_2019_us_tract_500k/cb_2019_us_tract_500k.shp")
us_tract_shp.head()

STATEFP COUNTYFP TRACTCE              AFFGEOID        GEOID     NAME LSAD  \
0      34      005  702808  1400000US34005702808  34005702808  7028.08   CT   
1      39      153  504200  1400000US39153504200  39153504200     5042   CT   
2      36      071  000502  1400000US36071000502  36071000502     5.02   CT   
3      06      013  370000  1400000US06013370000  06013370000     3700   CT   
4      48      113  019204  1400000US48113019204  48113019204   192.04   CT   

     ALAND  AWATER                                           geometry  
0  1510553       0  POLYGON ((-74.90801 40.03250, -74.90496 40.036...  
1   808151       0  POLYGON ((-81.51807 41.05957, -81.51805 41.062...  
2   967431  969216  POLYGON ((-74.02226 41.49281, -74.02180 41.496...  
3   999356       0  POLYGON ((-122.32697 37.93502, -122.32610 37.9...  
4  2851229    9966  POLYGON ((-96.76902 32.94655, -96.76899 32.950...

In [30]:
ny_tracts = us_tract_shp.loc[us_tract_shp['STATEFP']=='36'].to_crs(4326).reset_index(drop=True)
ny_tracts.head()

STATEFP COUNTYFP TRACTCE              AFFGEOID        GEOID     NAME LSAD  \
0      36      071  000502  1400000US36071000502  36071000502     5.02   CT   
1      36      103  135208  1400000US36103135208  36103135208  1352.08   CT   
2      36      047  057800  1400000US36047057800  36047057800      578   CT   
3      36      047  058900  1400000US36047058900  36047058900      589   CT   
4      36      055  013204  1400000US36055013204  36055013204   132.04   CT   

      ALAND  AWATER                                           geometry  
0    967431  969216  POLYGON ((-74.02226 41.49281, -74.02180 41.496...  
1   2287077       0  POLYGON ((-73.28263 40.83063, -73.28157 40.832...  
2    172233       0  POLYGON ((-73.95398 40.60140, -73.95304 40.601...  
3    424025   38353  POLYGON ((-73.94605 40.72926, -73.94419 40.729...  
4  28207247   93541  POLYGON ((-77.66835 43.02829, -77.66806 43.029...

In [31]:
fixed_tiles_ny = gpd.sjoin(fixed_ookla_shp, ny_tracts, how='inner', op='intersects')
fixed_ny = fixed_tiles_ny.reset_index(drop=True)
fixed_ny.head()

quadkey  avg_d_kbps  avg_u_kbps  avg_lat_ms  tests  devices  \
0  0320101110331012      112600       79229           8     99       21   
1  0320101110331011      130626       67490           9    118       12   
2  0320101110331003      125738       95820           7     22       12   
3  0320101110331020       98630       65159          11     16        8   
4  0320101110313232      129067      111693           8     75       18   

                                            geometry  index_right STATEFP  \
0  POLYGON ((-73.50952 40.77638, -73.50403 40.776...         1052      36   
1  POLYGON ((-73.50403 40.78054, -73.49854 40.780...         1052      36   
2  POLYGON ((-73.51501 40.77638, -73.50952 40.776...         1052      36   
3  POLYGON ((-73.52051 40.77222, -73.51501 40.772...         1052      36   
4  POLYGON ((-73.50952 40.78470, -73.50403 40.784...         1052      36   

  COUNTYFP TRACTCE              AFFGEOID        GEOID  NAME LSAD    ALAND  \
0      059  519400  1400000US36059519400  36059519400  5194   CT  1279432   
1      059  519400  1400000US36059519400  36059519400  5194   CT  1279432   
2      059  519400  1400000US36059519400  36059519400  5194   CT  1279432   
3      059  519400  1400000US36059519400  36059519400  5194   CT  1279432   
4      059  519400  1400000US36059519400  36059519400  5194   CT  1279432   

   AWATER  
0       0  
1       0  
2       0  
3       0  
4       0

In [32]:
mobile_tiles_ny = gpd.sjoin(mobile_ookla_shp, ny_tracts, how='inner', op='intersects')
mobile_ny = mobile_tiles_ny.reset_index(drop=True)
mobile_ny.head()

quadkey  avg_d_kbps  avg_u_kbps  avg_lat_ms  tests  devices  \
0  0302323301100220       25106        6284          56      1        1   
1  0302323301031300        3814        6047         109      2        2   
2  0302323301013330         913        7996          37      1        1   
3  0302323301013311       22793        5939          49      1        1   
4  0302323301031213        3581        5138          31      1        1   

                                            geometry  index_right STATEFP  \
0  POLYGON ((-74.00391 42.00849, -73.99841 42.008...         4746      36   
1  POLYGON ((-74.02588 41.95132, -74.02039 41.951...         4746      36   
2  POLYGON ((-74.01489 41.97583, -74.00940 41.975...         4746      36   
3  POLYGON ((-74.00940 41.98399, -74.00391 41.983...         4746      36   
4  POLYGON ((-74.03137 41.94723, -74.02588 41.947...         4746      36   

  COUNTYFP TRACTCE              AFFGEOID        GEOID  NAME LSAD     ALAND  \
0      111  951500  1400000US36111951500  36111951500  9515   CT  48023363   
1      111  951500  1400000US36111951500  36111951500  9515   CT  48023363   
2      111  951500  1400000US36111951500  36111951500  9515   CT  48023363   
3      111  951500  1400000US36111951500  36111951500  9515   CT  48023363   
4      111  951500  1400000US36111951500  36111951500  9515   CT  48023363   

   AWATER  
0  756736  
1  756736  
2  756736  
3  756736  
4  756736

In [33]:
fixed_ny.to_file('data/NY/fixed_ny.shp')
mobile_ny.to_file('data/NY/mobile_ny.shp')

/var/folders/0f/dj_hz1316w7c4qhdk33w3p3w0000gn/T/ipykernel_16767/4096207189.py:1: UserWarning: Column names longer than 10 characters will be truncated when saved to ESRI Shapefile.
  fixed_ny.to_file('data/NY/fixed_ny.shp')
/var/folders/0f/dj_hz1316w7c4qhdk33w3p3w0000gn/T/ipykernel_16767/4096207189.py:2: UserWarning: Column names longer than 10 characters will be truncated when saved to ESRI Shapefile.
  mobile_ny.to_file('data/NY/mobile_ny.shp')


In [34]:
# Check number of unique GEOID
mobile_ny.GEOID.value_counts()

36081071600    68
36059407301    46
36055013104    45
36059300900    41
36081038302    40
               ..
36059416402     1
36001001901     1
36037940100     1
36063024301     1
36065025800     1
Name: GEOID, Length: 4842, dtype: int64

In [35]:
# Check number of unique GEOIDs
fixed_ny.GEOID.value_counts()

36103190704    176
36103200902    170
36059517701    162
36069050102    122
36119012400    116
              ... 
36081990100      1
36081091602      1
36005017902      1
36009940000      1
36005003800      1
Name: GEOID, Length: 4894, dtype: int64

I need to incorporate Group 1A's code for weighted averages, but I aggregated the data by taking the average or sum depending on the column.

In [38]:
wm = lambda x: np.average(x, weights=mobile_ny.loc[x.index, 'tests'])

mobile_agg = mobile_ny.groupby(['GEOID'], as_index=False).agg(
                            m_d_kbps_wm = ('avg_d_kbps', wm),
                            m_u_kbps_wm = ('avg_u_kbps', wm),
                            m_tests = ('tests', sum),
                            m_devices = ('devices', sum)
                            )
mobile_agg.head()

GEOID   m_d_kbps_wm   m_u_kbps_wm  m_tests  m_devices
0  36001000100  24171.750000  11767.375000       16         13
1  36001000200  22492.375000  11001.193182       88         33
2  36001000300  33222.559524  12156.107143       84         40
3  36001000401  51093.853659  14892.902439       41         20
4  36001000403  40498.808824  10517.750000       68         36

In [47]:
type(mobile_agg.GEOID[0])

str

In [40]:
mobile_agg.shape

(4842, 5)

In [39]:
wm = lambda x: np.average(x, weights=fixed_ny.loc[x.index, 'tests'])

fixed_agg = fixed_ny.groupby(['GEOID'], as_index=False).agg(
                            f_d_kbps_wm = ('avg_d_kbps', wm),
                            f_u_kbps_wm = ('avg_u_kbps', wm),
                            f_tests = ('tests', sum),
                            f_devices = ('devices', sum)
                            )
fixed_agg.head()

GEOID    f_d_kbps_wm   f_u_kbps_wm  f_tests  f_devices
0  36001000100   65004.333333  10497.234568       81         28
1  36001000200   77159.277778  12049.351852      108         58
2  36001000300   84453.176923  22220.796154      260         95
3  36001000401   92439.055556  43385.986111       72         35
4  36001000403  102956.056980  13773.612536      351        115

In [41]:
fixed_agg.shape

(4894, 5)

In [46]:
type(fixed_agg.GEOID[0])

str

In [56]:
fixed_agg.GEOID.value_counts()

36001000100    1
36081018700    1
36081019800    1
36081019600    1
36081019400    1
              ..
36047067200    1
36047067000    1
36047066600    1
36047066200    1
36123150500    1
Name: GEOID, Length: 4894, dtype: int64

In [57]:
mobile_agg.GEOID.value_counts()

36001000100    1
36081018900    1
36081018700    1
36081018600    1
36081018502    1
              ..
36047068800    1
36047068600    1
36047068200    1
36047068000    1
36123150500    1
Name: GEOID, Length: 4842, dtype: int64

In [50]:
fixed_agg['GEOID'] = fixed_agg['GEOID'].apply(lambda x: int(x))
mobile_agg['GEOID'] = mobile_agg['GEOID'].apply(lambda x: int(x))

In [58]:
df_3 = mobile_agg.merge(fixed_agg, how='outer', on='GEOID')
df_3.head()

GEOID   m_d_kbps_wm   m_u_kbps_wm  m_tests  m_devices    f_d_kbps_wm  \
0  36001000100  24171.750000  11767.375000     16.0       13.0   65004.333333   
1  36001000200  22492.375000  11001.193182     88.0       33.0   77159.277778   
2  36001000300  33222.559524  12156.107143     84.0       40.0   84453.176923   
3  36001000401  51093.853659  14892.902439     41.0       20.0   92439.055556   
4  36001000403  40498.808824  10517.750000     68.0       36.0  102956.056980   

    f_u_kbps_wm  f_tests  f_devices  
0  10497.234568     81.0       28.0  
1  12049.351852    108.0       58.0  
2  22220.796154    260.0       95.0  
3  43385.986111     72.0       35.0  
4  13773.612536    351.0      115.0

In [59]:
df_3.shape

(4897, 9)

In [79]:
df_3['GEOID'] = df_3['GEOID'].apply(lambda x: str(x))
df_tot = df_2.merge(df_3, how='left', on='GEOID')
df_tot.head()

GEOID                                           geometry  \
0  36001000100  POLYGON ((-73.74506 42.67228, -73.74374 42.677...   
1  36001000200  POLYGON ((-73.75993 42.65997, -73.76156 42.660...   
2  36001000300  POLYGON ((-73.83021 42.69418, -73.82785 42.696...   
3  36001000401  POLYGON ((-73.89421 42.70977, -73.89668 42.711...   
4  36001000403  POLYGON ((-73.82108 42.67857, -73.81916 42.680...   

   f_broadband  f_computer   f_ba  f_black  f_hispanic  log_mhi      mhi  \
0        0.836       0.896  0.137    0.740       0.128    10.39  32389.0   
1        0.767       0.903  0.225    0.845       0.043    10.23  27714.0   
2        0.819       0.881  0.289    0.435       0.121    10.72  45272.0   
3        0.837       0.857  0.476    0.076       0.046    11.22  74274.0   
4        0.817       0.861  0.591    0.129       0.058    11.22  74426.0   

   population  households  broadband  computer  black  hispanic    mhi.1  \
0      2035.0       800.0       83.6      89.6   74.0      12.8  32389.0   
1      4793.0      2213.0       76.7      90.3   84.5       4.3  27714.0   
2      6147.0      2362.0       81.9      88.1   43.5      12.1  45272.0   
3      2362.0       997.0       83.7      85.7    7.6       4.6  74274.0   
4      4253.0      2186.0       81.7      86.1   12.9       5.8  74426.0   

     ba  den_computers  n_computer  n_broadband  den_black  n_black  \
0  13.7          800.0       717.0        669.0     2035.0   1506.0   
1  22.5         2213.0      1999.0       1697.0     4793.0   4050.0   
2  28.9         2362.0      2081.0       1935.0     6147.0   2674.0   
3  47.6          997.0       854.0        834.0     2362.0    180.0   
4  59.1         2186.0      1882.0       1786.0     4253.0    548.0   

   den_hispanic  n_hispanic  den_ba    n_ba  nhh_computer  nhh_broadband  \
0        2035.0       261.0  1283.0   176.0           717            669   
1        4793.0       206.0  2710.0   609.0          1999           1697   
2        6147.0       742.0  3725.0  1078.0          2081           1935   
3        2362.0       109.0  2169.0  1032.0           854            834   
4        4253.0       246.0  3220.0  1904.0          1882           1786   

   nhh_computer_any_internet  nhh_computer_and_dialup  \
0                        717                        0   
1                       1999                        0   
2                       2081                       11   
3                        854                        0   
4                       1882                        0   

   nhh_computer_and_broadband  nhh_computer_no_internet  nhh_no_computer  \
0                         669                        48               83   
1                        1685                       314              214   
2                        1917                       153              281   
3                         826                        28              143   
4                        1733                       149              304   

   den_age  n_children  n_children_computer  n_children_computer_and_dialup  \
0     2022         532                  520                               0   
1     4711        1274                 1221                               0   
2     5925        1729                 1717                               0   
3     1790         171                  171                               0   
4     4240         462                  462                               0   

   n_children_computer_and_broadband  n_children_computer_no_internet  \
0                                511                                9   
1                               1138                               83   
2                               1630                               87   
3                                171                                0   
4                                462                                0   

   n_children_no_computer      max_dn     max_up      dn10     dn100  \
0         

Need to look into joining tiles to the counties: 

```
tiles_in_ky_counties = gpd.sjoin(tiles, ky_counties, how='inner', op='intersects')
```

May account for the missing areas in this dataset.  The Ookla dataset we were provided is not a geodataframe, and sjoin only works on two geodataframes.  I managed to download a geo version of the ookla dataset, but have not incorporated yet into notebook.

In [80]:
df_tot.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
Int64Index: 4906 entries, 0 to 4905
Data columns (total 54 columns):
 #   Column                             Non-Null Count  Dtype   
---  ------                             --------------  -----   
 0   GEOID                              4906 non-null   object  
 1   geometry                           4906 non-null   geometry
 2   f_broadband                        4906 non-null   float64 
 3   f_computer                         4906 non-null   float64 
 4   f_ba                               4906 non-null   float64 
 5   f_black                            4906 non-null   float64 
 6   f_hispanic                         4906 non-null   float64 
 7   log_mhi                            4785 non-null   float64 
 8   mhi                                4906 non-null   float64 
 9   population                         4906 non-null   float64 
 10  households                         4906 non-null   float64 
 11  broadband                          

In [81]:
df_tot.isna().sum()

GEOID                                  0
geometry                               0
f_broadband                            0
f_computer                             0
f_ba                                   0
f_black                                0
f_hispanic                             0
log_mhi                              121
mhi                                    0
population                             0
households                             0
broadband                              0
computer                               0
black                                  0
hispanic                               0
mhi.1                                  0
ba                                     0
den_computers                          0
n_computer                             0
n_broadband                            0
den_black                              0
n_black                                0
den_hispanic                           0
n_hispanic                             0
den_ba          

In [82]:
df_tot.to_file("data/NY/ny_combined.shp")

/var/folders/0f/dj_hz1316w7c4qhdk33w3p3w0000gn/T/ipykernel_16767/1179071031.py:1: UserWarning: Column names longer than 10 characters will be truncated when saved to ESRI Shapefile.
  df_tot.to_file("data/NY/ny_combined.shp")


# Illinois Datasets

In [85]:
# shapefile = "/content/drive/MyDrive/Datasets/Broadband/IL/il_spdf/il_spdf.shp"
# acsfile = '/content/drive/MyDrive/Datasets/Broadband/IL/acs_2019_IL.csv'
# fccfile = '/content/drive/MyDrive/Datasets/Broadband/IL/fcc_477_census_tract_IL.csv'
# ooklafile = '/content/drive/MyDrive/Datasets/Broadband/IL/ookla_combined_il.csv'

shapefile = "data/IL/il_spdf/il_spdf.shp"
acsfile = 'data/IL/acs_2019_IL.csv'
fccfile = 'data/IL/fcc_477_census_tract_IL.csv'
ooklafile = 'data/IL/ookla_combined_il.csv'

In [86]:
shpfile = gpd.read_file(shapefile)
shpfile = shpfile.sort_values(by=['GEOID'])
shpfile.head()

STATEFP COUNTYFP TRACTCE              AFFGEOID        GEOID  NAME LSAD  \
1992      17      001  000100  1400000US17001000100  17001000100     1   CT   
1980      17      001  000201  1400000US17001000201  17001000201  2.01   CT   
22        17      001  000202  1400000US17001000202  17001000202  2.02   CT   
1117      17      001  000400  1400000US17001000400  17001000400     4   CT   
784       17      001  000500  1400000US17001000500  17001000500     5   CT   

           ALAND    AWATER                                           geometry  
1992  12937185.0   22042.0  POLYGON ((-91.37766 39.94160, -91.37759 39.946...  
1980   3206335.0       0.0  POLYGON ((-91.39646 39.95621, -91.39631 39.965...  
22     1514956.0       0.0  POLYGON ((-91.39370 39.94678, -91.37759 39.946...  
1117   4458010.0  211712.0  POLYGON ((-91.42005 39.95081, -91.41917 39.951...  
784    1262517.0       0.0  POLYGON ((-91.40340 39.95048, -91.39655 39.950...

In [87]:
data = pd.read_csv(acsfile)
data = data.rename(columns={'geoid':'GEOID'})
data['GEOID'] = data['GEOID'].apply(lambda x: str(x))
data.drop(columns=['state', 'county', 'tract', 'state_lkp'], inplace=True)
data.head()

GEOID  f_broadband  f_computer   f_ba  f_black  f_hispanic  log_mhi  \
0  17001000100        0.813       0.861  0.246    0.068       0.000    10.78   
1  17001000201        0.879       0.940  0.266    0.078       0.016    10.71   
2  17001000202        0.863       0.933  0.221    0.080       0.008    10.83   
3  17001000400        0.691       0.780  0.094    0.182       0.006    10.31   
4  17001000500        0.665       0.787  0.113    0.133       0.035    10.62   

       mhi  population  households  broadband  computer  black  hispanic  \
0  48088.0      4761.0      2202.0       81.3      86.1    6.8       0.0   
1  44855.0      1985.0       889.0       87.9      94.0    7.8       1.6   
2  50375.0      2376.0       970.0       86.3      93.3    8.0       0.8   
3  30164.0      3422.0      1316.0       69.1      78.0   18.2       0.6   
4  41008.0      2175.0       813.0       66.5      78.7   13.3       3.5   

     mhi.1    ba  den_computers  n_computer  n_broadband  den_black  n_black  \
0  48088.0  24.6         2202.0      1896.0       1791.0     4761.0    324.0   
1  44855.0  26.6          889.0       836.0        781.0     1985.0    155.0   
2  50375.0  22.1          970.0       905.0        837.0     2376.0    191.0   
3  30164.0   9.4         1316.0      1027.0        909.0     3422.0    624.0   
4  41008.0  11.3          813.0       640.0        541.0     2175.0    290.0   

   den_hispanic  n_hispanic  den_ba   n_ba  nhh_computer  nhh_broadband  \
0        4761.0         0.0  3381.0  831.0          1896           1791   
1        1985.0        32.0  1400.0  373.0           836            781   
2        2376.0        19.0  1527.0  338.0           905            837   
3        3422.0        20.0  2536.0  238.0          1027            909   
4        2175.0        77.0  1455.0  165.0           640            541   

   nhh_computer_any_internet  nhh_computer_and_dialup  \
0                       1896                        0   
1                        836                        5   
2                        905                        7   
3                       1027                        9   
4                        640                        0   

   nhh_computer_and_broadband  nhh_computer_no_internet  nhh_no_computer  \
0                        1791                       105              306   
1                         781                        50               53   
2                         829                        69               65   
3                         909                       109              289   
4                         536                       104              173   

   den_age  n_children  n_children_computer  n_children_computer_and_dialup  \
0     4741        1044                 1044                               0   
1     1961         383                  383                               0   
2     2336         521                  521                               0   
3     3147         645                  637                               0   
4     1878         475                  458                               0   

   n_children_computer_and_broadband  n_children_computer_no_internet  \
0                                978                               66   
1                                383                                0   
2                                498                               23   
3                                619                               18   
4                                434                               24   

   n_children_no_computer  
0                       0  
1                       0  
2                       0  
3                       8  
4                      17

In [88]:
data2 = pd.read_csv(fccfile)
data2 = data2.rename(columns={'tract':'GEOID'})
data2['GEOID'] = data2['GEOID'].apply(lambda x: str(x))
data2.drop(columns=['state'], inplace=True)
data2.head()

GEOID       max_dn       max_up      dn10     dn100     dn250  \
0  17001000100   931.548571   915.160000  2.268571  1.502857  1.502857   
1  17001000201   872.587302   796.936508  2.761905  1.603175  1.603175   
2  17001000202  1000.000000  1000.000000  3.944444  2.018519  2.018519   
3  17001000400   519.804124   198.982021  1.757732  0.680412  0.680412   
4  17001000500   982.120690   916.275862  3.775862  1.793103  1.793103   

   fiber_100u  
0    0.920000  
1    0.809524  
2    1.018519  
3    0.190722  
4    0.913793

In [83]:
il_tracts = us_tract_shp.loc[us_tract_shp['STATEFP']=='17'].to_crs(4326).reset_index(drop=True)
fixed_tiles_il = gpd.sjoin(fixed_ookla_shp, il_tracts, how='inner', op='intersects')
fixed_il = fixed_tiles_il.reset_index(drop=True)
mobile_tiles_il = gpd.sjoin(mobile_ookla_shp, il_tracts, how='inner', op='intersects')
mobile_il = mobile_tiles_il.reset_index(drop=True)

wm = lambda x: np.average(x, weights=mobile_il.loc[x.index, 'tests'])
mobile_agg = mobile_il.groupby(['GEOID'], as_index=False).agg(
                            m_d_kbps_wm = ('avg_d_kbps', wm),
                            m_u_kbps_wm = ('avg_u_kbps', wm),
                            m_tests = ('tests', sum),
                            m_devices = ('devices', sum)
                            )
wm = lambda x: np.average(x, weights=fixed_il.loc[x.index, 'tests'])
fixed_agg = fixed_il.groupby(['GEOID'], as_index=False).agg(
                            f_d_kbps_wm = ('avg_d_kbps', wm),
                            f_u_kbps_wm = ('avg_u_kbps', wm),
                            f_tests = ('tests', sum),
                            f_devices = ('devices', sum)
                            )
fixed_agg['GEOID'] = fixed_agg['GEOID'].apply(lambda x: int(x))
mobile_agg['GEOID'] = mobile_agg['GEOID'].apply(lambda x: int(x))
data3 = mobile_agg.merge(fixed_agg, how='outer', on='GEOID')
data3['GEOID'] = data3['GEOID'].apply(lambda x: str(x))

In [89]:
df_1 = shpfile.merge(data, how="left", on="GEOID")
df_2 = df_1.merge(data2, how="left", on="GEOID")
df_final = df_2.merge(data3, how="left", on="GEOID")
df_final.head()

STATEFP COUNTYFP TRACTCE              AFFGEOID        GEOID  NAME LSAD  \
0      17      001  000100  1400000US17001000100  17001000100     1   CT   
1      17      001  000201  1400000US17001000201  17001000201  2.01   CT   
2      17      001  000202  1400000US17001000202  17001000202  2.02   CT   
3      17      001  000400  1400000US17001000400  17001000400     4   CT   
4      17      001  000500  1400000US17001000500  17001000500     5   CT   

        ALAND    AWATER                                           geometry  \
0  12937185.0   22042.0  POLYGON ((-91.37766 39.94160, -91.37759 39.946...   
1   3206335.0       0.0  POLYGON ((-91.39646 39.95621, -91.39631 39.965...   
2   1514956.0       0.0  POLYGON ((-91.39370 39.94678, -91.37759 39.946...   
3   4458010.0  211712.0  POLYGON ((-91.42005 39.95081, -91.41917 39.951...   
4   1262517.0       0.0  POLYGON ((-91.40340 39.95048, -91.39655 39.950...   

   f_broadband  f_computer   f_ba  f_black  f_hispanic  log_mhi      mhi  \
0        0.813       0.861  0.246    0.068       0.000    10.78  48088.0   
1        0.879       0.940  0.266    0.078       0.016    10.71  44855.0   
2        0.863       0.933  0.221    0.080       0.008    10.83  50375.0   
3        0.691       0.780  0.094    0.182       0.006    10.31  30164.0   
4        0.665       0.787  0.113    0.133       0.035    10.62  41008.0   

   population  households  broadband  computer  black  hispanic    mhi.1  \
0      4761.0      2202.0       81.3      86.1    6.8       0.0  48088.0   
1      1985.0       889.0       87.9      94.0    7.8       1.6  44855.0   
2      2376.0       970.0       86.3      93.3    8.0       0.8  50375.0   
3      3422.0      1316.0       69.1      78.0   18.2       0.6  30164.0   
4      2175.0       813.0       66.5      78.7   13.3       3.5  41008.0   

     ba  den_computers  n_computer  n_broadband  den_black  n_black  \
0  24.6         2202.0      1896.0       1791.0     4761.0    324.0   
1  26.6          889.0       836.0        781.0     1985.0    155.0   
2  22.1          970.0       905.0        837.0     2376.0    191.0   
3   9.4         1316.0      1027.0        909.0     3422.0    624.0   
4  11.3          813.0       640.0        541.0     2175.0    290.0   

   den_hispanic  n_hispanic  den_ba   n_ba  nhh_computer  nhh_broadband  \
0        4761.0         0.0  3381.0  831.0          1896           1791   
1        1985.0        32.0  1400.0  373.0           836            781   
2        2376.0        19.0  1527.0  338.0           905            837   
3        3422.0        20.0  2536.0  238.0          1027            909   
4        2175.0        77.0  1455.0  165.0           640            541   

   nhh_computer_any_internet  nhh_computer_and_dialup  \
0                       1896                        0   
1                        836                        5   
2                        905                        7   
3                       1027                        9   
4                        640                        0   

   nhh_computer_and_broadband  nhh_computer_no_internet  nhh_no_computer  \
0                        1791                       105              306   
1                         781                        50               53   
2                         829                        69               65   
3                         909                       109              289   
4                         536                       104              173   

   den_age  n_children  n_children_computer  n_children_computer_and_dialup  \
0     4741        1044                 1044                               0   
1     1961         383                  383                               0   
2     2336         521                  521                               0   
3     3147         645                  637                               0   
4     1878         475                  458                               0   

   n_chil

In [90]:
df_final.to_file('data/IL/il_combined.shp')

/var/folders/0f/dj_hz1316w7c4qhdk33w3p3w0000gn/T/ipykernel_16767/2041146939.py:1: UserWarning: Column names longer than 10 characters will be truncated when saved to ESRI Shapefile.
  df_final.to_file('data/il_combined.shp')


# Formula to create dataset

In [91]:
# shapefile = "/content/drive/MyDrive/Datasets/Broadband/IL/il_spdf/il_spdf.shp"
# acsfile = '/content/drive/MyDrive/Datasets/Broadband/IL/acs_2019_IL.csv'
# fccfile = '/content/drive/MyDrive/Datasets/Broadband/IL/fcc_477_census_tract_IL.csv'
# ooklafile = '/content/drive/MyDrive/Datasets/Broadband/IL/ookla_combined_il.csv'
# FIPS=17


In [92]:
fixed_ookla = gpd.read_file('data/z-archive/Ookla/ookla-data/2019-01-01_performance_fixed_tiles/gps_fixed_tiles.shp')
mobile_ookla = gpd.read_file('data/z-archive/Ookla/ookla-data/2019-01-01_performance_mobile_tiles/gps_mobile_tiles.shp')
us_tract_shp = gpd.read_file("data/z-archive/Census Tract Shapefiles/cb_2019_us_tract_500k/cb_2019_us_tract_500k.shp")

In [95]:
us_tract_shp.STATEFP.value_counts()

06    8041
48    5254
36    4906
12    4210
42    3217
17    3121
39    2948
26    2772
37    2192
34    2004
13    1966
51    1900
04    1526
18    1508
47    1497
25    1474
53    1454
55    1396
24    1396
29    1393
27    1336
08    1249
01    1180
22    1140
21    1115
45    1103
40    1046
72     908
09     829
41     828
19     825
20     770
32     687
05     686
28     661
49     588
31     532
35     499
54     484
23     355
15     331
16     298
33     294
30     271
44     243
46     222
10     217
38     205
50     184
11     179
02     167
56     132
66      57
78      31
69      23
60      18
Name: STATEFP, dtype: int64

In [101]:
il_tracts = us_tract_shp.loc[us_tract_shp['STATEFP']=='06'].to_crs(4326).reset_index(drop=True)
fixed_tiles_il = gpd.sjoin(fixed_ookla_shp, il_tracts, how='inner', op='intersects')
fixed_il = fixed_tiles_il.reset_index(drop=True)
mobile_tiles_il = gpd.sjoin(mobile_ookla_shp, il_tracts, how='inner', op='intersects')
mobile_il = mobile_tiles_il.reset_index(drop=True)

wm = lambda x: np.average(x, weights=mobile_il.loc[x.index, 'tests'])
mobile_agg = mobile_il.groupby(['GEOID'], as_index=False).agg(
                            m_d_kbps_wm = ('avg_d_kbps', wm),
                            m_u_kbps_wm = ('avg_u_kbps', wm),
                            m_tests = ('tests', sum),
                            m_devices = ('devices', sum)
                            )
wm = lambda x: np.average(x, weights=fixed_il.loc[x.index, 'tests'])
fixed_agg = fixed_il.groupby(['GEOID'], as_index=False).agg(
                            f_d_kbps_wm = ('avg_d_kbps', wm),
                            f_u_kbps_wm = ('avg_u_kbps', wm),
                            f_tests = ('tests', sum),
                            f_devices = ('devices', sum)
                            )
fixed_agg['GEOID'] = fixed_agg['GEOID'].apply(lambda x: int(x))
mobile_agg['GEOID'] = mobile_agg['GEOID'].apply(lambda x: int(x))
data3 = mobile_agg.merge(fixed_agg, how='outer', on='GEOID')
data3['GEOID'] = data3['GEOID'].apply(lambda x: str(x))

In [100]:
# need to revise

def prepare_dataset(shapefile, acsfile, fccfile):
    shpfile = gpd.read_file(shapefile)
    shpfile = shpfile.sort_values(by=['GEOID'])
    data = pd.read_csv(acsfile)
    data = data.rename(columns={'geoid':'GEOID'})
    data['GEOID'] = data['GEOID'].apply(lambda x: str(x))
    data.drop(columns=['state', 'county', 'tract', 'state_lkp'], inplace=True)
    data2 = pd.read_csv(fccfile)
    data2 = data2.rename(columns={'tract':'GEOID'})
    data2['GEOID'] = data2['GEOID'].apply(lambda x: str(x))
    data2.drop(columns=['state'], inplace=True)
    tracts = us_tract_shp.loc[us_tract_shp['STATEFP']=='FIPS'].to_crs(4326).reset_index(drop=True)
    fixed_tiles = gpd.sjoin(fixed_ookla, tracts, how='inner', op='intersects')
    fixed = fixed_tiles.reset_index(drop=True)
    mobile_tiles = gpd.sjoin(mobile_ookla, tracts, how='inner', op='intersects')
    mobile = mobile_tiles.reset_index(drop=True)
    wm = lambda x: np.average(x, weights=mobile.loc[x.index, 'tests'])
    mobile_agg = mobile.groupby(['GEOID'], as_index=False).agg(
                            m_d_kbps_wm = ('avg_d_kbps', wm),
                            m_u_kbps_wm = ('avg_u_kbps', wm),
                            m_tests = ('tests', sum),
                            m_devices = ('devices', sum)
                            )
    wm = lambda x: np.average(x, weights=fixed.loc[x.index, 'tests'])
    fixed_agg = fixed.groupby(['GEOID'], as_index=False).agg(
                            f_d_kbps_wm = ('avg_d_kbps', wm),
                            f_u_kbps_wm = ('avg_u_kbps', wm),
                            f_tests = ('tests', sum),
                            f_devices = ('devices', sum)
                            )
    fixed_agg['GEOID'] = fixed_agg['GEOID'].apply(lambda x: int(x))
    mobile_agg['GEOID'] = mobile_agg['GEOID'].apply(lambda x: int(x))
    data3 = mobile_agg.merge(fixed_agg, how='outer', on='GEOID')
    data3['GEOID'] = data3['GEOID'].apply(lambda x: str(x))
    df_1 = shpfile.merge(data, how="left", on="GEOID")
    df_2 = df_1.merge(data2, how="left", on="GEOID")
    df_final = df_2.merge(data3, how="left", on="GEOID")
    return df_final


In [17]:
il_combined = prepare_dataset(shapefile, acsfile, fccfile, ooklafile)
il_combined

STATEFP COUNTYFP TRACTCE              AFFGEOID        GEOID    NAME LSAD  \
0         17      001  000100  1400000US17001000100  17001000100       1   CT   
1         17      001  000201  1400000US17001000201  17001000201    2.01   CT   
2         17      001  000202  1400000US17001000202  17001000202    2.02   CT   
3         17      001  000400  1400000US17001000400  17001000400       4   CT   
4         17      001  000500  1400000US17001000500  17001000500       5   CT   
...      ...      ...     ...                   ...          ...     ...  ...   
3116      17      203  030501  1400000US17203030501  17203030501  305.01   CT   
3117      17      203  030502  1400000US17203030502  17203030502  305.02   CT   
3118      17      203  030601  1400000US17203030601  17203030601  306.01   CT   
3119      17      203  030602  1400000US17203030602  17203030602  306.02   CT   
3120      17      203  030700  1400000US17203030700  17203030700     307   CT   

            ALAND      AWATER  \
0      12937185.0     22042.0   
1       3206335.0         0.0   
2       1514956.0         0.0   
3       4458010.0    211712.0   
4       1262517.0         0.0   
...           ...         ...   
3116   44516795.0    111544.0   
3117   35324557.0  22436047.0   
3118  107361660.0    157898.0   
3119  191120324.0   1123588.0   
3120  235145636.0    120371.0   

                                               geometry  f_broadband  \
0     POLYGON ((-91.37766 39.94160, -91.37759 39.946...        0.813   
1     POLYGON ((-91.39646 39.95621, -91.39631 39.965...        0.879   
2     POLYGON ((-91.39370 39.94678, -91.37759 39.946...        0.863   
3     POLYGON ((-91.42005 39.95081, -91.41917 39.951...        0.691   
4     POLYGON ((-91.40340 39.95048, -91.39655 39.950...        0.665   
...                                                 ...          ...   
3116  POLYGON ((-89.51271 40.76214, -89.51124 40.769...        0.886   
3117  POLYGON ((-89.56252 40.79285, -89.56239 40.793...        0.833   
3118  POLYGON ((-89.33009 40.74826, -89.31363 40.748...        0.793   
3119  POLYGON ((-89.32869 40.66078, -89.31987 40.660...        0.816   
3120  POLYGON ((-89.17682 40.75731, -89.15773 40.757...        0.754   

      f_computer   f_ba  f_black  f_hispanic  log_mhi       mhi  population  \
0          0.861  0.246    0.068       0.000    10.78   48088.0      4761.0   
1          0.940  0.266    0.078       0.016    10.71   44855.0      1985.0   
2          0.933  0.221    0.080       0.008    10.83   50375.0      2376.0   
3          0.780  0.094    0.182       0.006    10.31   30164.0      3422.0   
4          0.787  0.113    0.133       0.035    10.62   41008.0      2175.0   
...          ...    ...      ...         ...      ...       ...         ...   
3116       0.942  0.526    0.014       0.017    11.52  100875.0      7795.0   
3117       0.906  0.168    0.002       0.015    11.02   61293.0      2500.0   
3118       0.876  0.384    0.020       0.034    11.00   59929.0      6620.0   
3119       0.922  0.327    0.013       0.005    11.40   89650.0      3437.0   
3120       0.868  0.237    0.025       0.005    11.20   73421.0      4226.0   

      households  broadband  computer  black  hispanic     mhi.1    ba  \
0         2202.0       81.3      86.1    6.8       0.0   48088.0  24.6   
1          889.0       87.9      94.0    7.8       1.6   44855.0  26.6   
2          970.0       86.3      93.3    8.0       0.8   50375.0  22.1   
3         1316.0       69.1      78.0   18.2       0.6   30164.0   9.4   
4          813.0       66.5      78.7   13.3       3.5   41008.0  11.3   
...          ...        ...       ...    ...       ...       ...   ...   
3116      2758.0       88.6      94.2    1.4       1.7  100875.0  52.6   
3117      1084.0       83.3      90.6    0.2       1.5   61293.0  16.8   
3118      2371.0       79.3      87.6    2.0       3.4   59929.0  38.4   
3119      1261.0       81.6      92.2    1.3       0.5   89650.0  32.7   
3120   

## California Dataset

In [7]:
shapefile = "/content/drive/MyDrive/Datasets/Broadband/CA/ca_spdf/ca_spdf.shp"
acsfile = '/content/drive/MyDrive/Datasets/Broadband/CA/acs_2019_CA.csv'
fccfile = '/content/drive/MyDrive/Datasets/Broadband/CA/fcc_477_census_tract_CA.csv'
ooklafile = '/content/drive/MyDrive/Datasets/Broadband/CA/ookla_combined_ca.csv'

In [94]:
shapefile = "data/CA/ca_spdf/ca_spdf.shp"
acsfile = 'data/CA/acs_2019_CA.csv'
fccfile = 'data/CA/fcc_477_census_tract_CA.csv'
ooklafile = 'data/CA/ookla_combined_ca.csv'
FIPS=06

In [99]:
ca_df = prepare_dataset(shapefile, acsfile, fccfile, ooklafile)
ca_df

STATEFP COUNTYFP TRACTCE              AFFGEOID        GEOID    NAME LSAD  \
0         06      001  400100  1400000US06001400100  06001400100    4001   CT   
1         06      001  400200  1400000US06001400200  06001400200    4002   CT   
2         06      001  400300  1400000US06001400300  06001400300    4003   CT   
3         06      001  400400  1400000US06001400400  06001400400    4004   CT   
4         06      001  400500  1400000US06001400500  06001400500    4005   CT   
...      ...      ...     ...                   ...          ...     ...  ...   
8036      06      115  040800  1400000US06115040800  06115040800     408   CT   
8037      06      115  040901  1400000US06115040901  06115040901  409.01   CT   
8038      06      115  040902  1400000US06115040902  06115040902  409.02   CT   
8039      06      115  041000  1400000US06115041000  06115041000     410   CT   
8040      06      115  041100  1400000US06115041100  06115041100     411   CT   

            ALAND      AWATER  \
0       6894339.0         0.0   
1        586559.0         0.0   
2       1105851.0         0.0   
3        715630.0         0.0   
4        590297.0         0.0   
...           ...         ...   
8036  109401417.0    128316.0   
8037  263652570.0   4473094.0   
8038   93617079.0    240059.0   
8039  468397390.0   6927408.0   
8040  559316286.0  17080393.0   

                                               geometry  f_broadband  \
0     POLYGON ((-122.24692 37.88544, -122.24197 37.8...          NaN   
1     POLYGON ((-122.25742 37.84310, -122.25620 37.8...          NaN   
2     POLYGON ((-122.26416 37.84000, -122.26186 37.8...          NaN   
3     POLYGON ((-122.26180 37.84179, -122.26130 37.8...          NaN   
4     POLYGON ((-122.26941 37.84811, -122.26841 37.8...          NaN   
...                                                 ...          ...   
8036  POLYGON ((-121.51583 39.03013, -121.51504 39.0...          NaN   
8037  POLYGON ((-121.58249 39.13653, -121.58036 39.1...          NaN   
8038  POLYGON ((-121.47709 39.14366, -121.46775 39.1...          NaN   
8039  POLYGON ((-121.63631 39.24941, -121.63581 39.2...          NaN   
8040  POLYGON ((-121.37257 39.37358, -121.37048 39.3...          NaN   

      f_computer  f_ba  f_black  f_hispanic  log_mhi  mhi  population  \
0            NaN   NaN      NaN         NaN      NaN  NaN         NaN   
1            NaN   NaN      NaN         NaN      NaN  NaN         NaN   
2            NaN   NaN      NaN         NaN      NaN  NaN         NaN   
3            NaN   NaN      NaN         NaN      NaN  NaN         NaN   
4            NaN   NaN      NaN         NaN      NaN  NaN         NaN   
...          ...   ...      ...         ...      ...  ...         ...   
8036         NaN   NaN      NaN         NaN      NaN  NaN         NaN   
8037         NaN   NaN      NaN         NaN      NaN  NaN         NaN   
8038         NaN   NaN      NaN         NaN      NaN  NaN         NaN   
8039         NaN   NaN      NaN         NaN      NaN  NaN         NaN   
8040         NaN   NaN      NaN         NaN      NaN  NaN         NaN   

      households  broadband  computer  black  hispanic  mhi.1  ba  \
0            NaN        NaN       NaN    NaN       NaN    NaN NaN   
1            NaN        NaN       NaN    NaN       NaN    NaN NaN   
2            NaN        NaN       NaN    NaN       NaN    NaN NaN   
3            NaN        NaN       NaN    NaN       NaN    NaN NaN   
4            NaN        NaN       NaN    NaN       NaN    NaN NaN   
...          ...        ...       ...    ...       ...    ...  ..   
8036         NaN        NaN       NaN    NaN       NaN    NaN NaN   
8037         NaN        NaN       NaN    NaN       NaN    NaN NaN   
8038         NaN        NaN       NaN    NaN       NaN    NaN NaN   
8039         NaN        NaN       NaN    NaN       NaN    NaN NaN   
8040         NaN        NaN       NaN    NaN       NaN    NaN NaN   

      den_computers  n_computer  n_broadband  den_black  n_black